feat(i18n): translate src/content/learn/managing-state.md from English to Vietnamese (#476)
reactjs/vi.reactjs.org: src/content/learn/managing-state.md

---
title: Quản Lý State
---

<Intro>

> Khi ứng dụng của bạn ngày càng lớn, việc có ý thức hơn về cách tổ chức state và luồng data chảy giữa các component sẽ rất hữu ích. State dư thừa hoặc trùng lặp là nguyên nhân phổ biến gây ra một loạt bug khó hiểu. Trong chương này, bạn sẽ học cách cấu trúc state một cách tốt, cách giữ cho logic cập nhật state dễ bảo trì, và cách chia sẻ state giữa các component ở xa nhau.

Reviewer: ```suggestion
Khi ứng dụng của bạn ngày càng lớn, việc có chủ ý hơn về cách tổ chức state và luồng data chảy giữa các component sẽ rất hữu ích. State dư thừa hoặc trùng lặp là nguyên nhân phổ biến gây ra một loạt bug khó hiểu. Trong chương này, bạn sẽ học cách cấu trúc state một cách chặt chẽ, cách giữ cho logic cập nhật state dễ bảo trì, và cách chia sẻ state giữa các component ở xa nhau.
```

</Intro>

<YouWillLearn isChapter={true}>

> * [Cách suy nghĩ về thay đổi UI như những thay đổi state](/learn/reacting-to-input-with-state)

Reviewer: ```suggestion
* [Cách suy nghĩ về thay đổi UI khi state thay đổi](/learn/reacting-to-input-with-state)
```

> * [Cách cấu trúc state một cách tốt](/learn/choosing-the-state-structure)

Reviewer: ```suggestion
* [Cách cấu trúc state một cách chặt chẽ](/learn/choosing-the-state-structure)
```

<LearnMore path="/learn/reacting-to-input-with-state">

Đọc **[Phản Hồi Đầu Vào Thông Qua State](/learn/reacting-to-input-with-state)** để học cách tiếp cận các tương tác với mô hình tư duy điều khiển bằng state.

</LearnMore>

## Chọn cấu trúc state {/*choosing-the-state-structure*/}

> Cấu trúc state tốt có thể tạo ra sự khác biệt giữa một component dễ dàng sửa đổi và debug, và một component có hành vi khó đoán với một loạt bug khó hiểu. Nguyên tắc quan trọng nhất là state không nên chứa thông tin dư thừa hoặc trùng lặp. Nếu có state không cần thiết, rất dễ quên cập nhật nó và tạo ra bug!

Reviewer: ```suggestion
Cấu trúc state tốt có thể tạo ra sự khác biệt giữa một component dễ dàng sửa đổi và debug, và một component có hành vi khó đoán với một loạt bug khó hiểu. Nguyên tắc quan trọng nhất là state không nên chứa thông tin dư thừa hoặc trùng lặp. Nếu có state không cần thiết, ta rất dễ quên cập nhật nó và tạo ra bug!
```

</Sandpack>

Điều này có vẻ như một thay đổi nhỏ, nhưng nhiều bug khó hiểu trong các ứng dụng React được sửa theo cách này.

<LearnMore path="/learn/choosing-the-state-structure">

> Đọc **[Chọn Cấu Trúc State](/learn/choosing-the-state-structure)** để học cách thiết kế hình dạng state để tránh bug.

Reviewer: ```suggestion
Đọc **[Chọn Cấu Trúc State](/learn/choosing-the-state-structure)** để học cách thiết kế cấu trúc state để tránh bug.
```

<LearnMore path="/learn/extracting-state-logic-into-a-reducer">

Đọc **[Trích Xuất Logic State Vào Reducer](/learn/extracting-state-logic-into-a-reducer)** để học cách hợp nhất logic trong function reducer.

</LearnMore>

## Truyền data sâu với context {/*passing-data-deeply-with-context*/}

> Thông thường, bạn sẽ truyền thông tin từ component cha đến component con thông qua props. Nhưng việc truyền props có thể trở nên bất tiện nếu bạn cần truyền một prop qua nhiều component, hoặc nếu nhiều component cần cùng một thông tin. Context cho phép component cha cung cấp một số thông tin cho bất kỳ component nào trong cây con bên dưới nó—bất kể nó sâu đến đâu—mà không cần truyền một cách rõ ràng thông qua props.

Reviewer: ```suggestion
Thông thường, bạn sẽ truyền thông tin từ component cha đến component con thông qua props. Nhưng việc truyền props có thể trở nên bất tiện nếu bạn cần truyền một prop qua nhiều component, hoặc nếu nhiều component cần cùng một thông tin. Context cho phép component cha cung cấp một số thông tin cho bất kỳ component nào ở bên dưới nó trong cấu trúc cây—bất kể nó sâu đến đâu—mà không cần truyền một cách rõ ràng thông qua props.
```

<LearnMore path="/learn/passing-data-deeply-with-context">

Đọc **[Truyền Data Sâu Với Context](/learn/passing-data-deeply-with-context)** để học về việc sử dụng context như một giải pháp thay thế cho việc truyền props.

</LearnMore>

## Mở rộng với reducer và context {/*scaling-up-with-reducer-and-context*/}

> Reducer cho phép bạn tổng hợp logic cập nhật state của component. Context cho phép bạn truyền thông tin xuống sâu cho các component khác. Bạn có thể kết hợp reducer và context cùng nhau để quản lý state của một màn hình phức tạp.

Reviewer: ```suggestion
Reducer cho phép bạn tổng hợp logic cập nhật state của component. Context cho phép bạn truyền thông tin xuống sâu cho các component khác. Bạn có thể kết hợp reducer và context cùng nhau để quản lý state của một giao diện phức tạp.
```

> Với cách tiếp cận này, một component cha có state phức tạp sẽ quản lý nó bằng reducer. Các component khác ở bất kỳ vị trí sâu nào trong cây có thể đọc state của nó thông qua context. Chúng cũng có thể dispatch các action để cập nhật state đó.

Reviewer: ```suggestion
Với cách tiếp cận này, một component cha có state phức tạp sẽ quản lý nó bằng reducer. Các component khác ở bất kỳ vị trí nào sâu trong cây đều có thể đọc state của nó thông qua context. Chúng cũng có thể dispatch các action để cập nhật state đó.
```
